### ISSUE TITLE
CMake target naming convention

### ISSUE THREAD

**Issue by jngrad** [espressomd/espresso#4524]
**TL;DR:** ESPResSo uses CMake target names prone to collisions with the external CMake projects it imports. This can be remediated by adding the prefix `Espresso` to all target names. Using a namespace for target names would further help diagnosing CMake errors.

## Background on CMake targets

When including an external library, either through `FetchContent` or `find_package`, all targets of the imported library are included in the project global namespace, which can lead to name collision. For example, if a library declares a target `core`, the ESPResSo project cannot declare a target `core`, otherwise the following CMake error is raised:

```
CMake Error at CMakeLists.txt:14 (add_library):
  add_library cannot create target "core" because another target with the
  same name already exists.  The existing target is a shared library created
  in source directory
  "/work/jgrad/espresso/build/_deps/walberla-src".  See documentation
  for policy CMP0002 for more details.
```

To avoid this issue, projects designed to be packaged as pre-built shared objects use the following pattern:

```cmake
cmake_minimum_required(VERSION 3.16)
project(mylibrary VERSION 1.0.0 LANGUAGES CXX)

add_library(core SHARED core.cpp)
install(TARGETS core EXPORT MylibraryTargets LIBRARY DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
install(EXPORT MylibraryTargets NAMESPACE Mylibrary:: DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
export(EXPORT MylibraryTargets NAMESPACE Mylibrary::) # non-install version
```

This creates both the shared object `libcore.so` as well as a file `FindMylibrary.cmake` that contains the following logic:

```cmake
# Create imported target Mylibrary::core
add_library(Mylibrary::core SHARED IMPORTED)

# Import target "Mylibrary::core" for configuration ""
set_property(TARGET Mylibrary::core APPEND PROPERTY IMPORTED_CONFIGURATIONS NOCONFIG)
set_target_properties(Mylibrary::core PROPERTIES
  IMPORTED_LOCATION_NOCONFIG "/usr/lib/libcore.so"
  IMPORTED_SONAME_NOCONFIG "libcore.so"
)
```

One then uses `find_package(Mylibrary 1.0.0 REQUIRED core)` to load the absolute path to the shared object in a namespaced CMake target `Mylibrary::core`. Note that no `core` target is created by `find_package()`, thus allowing the consumer project to declare a `core` target. This is the solution used by the Boost library, for example.

For projects imported via `FetchContent`, the following pattern is sometimes used:

```cmake
cmake_minimum_required(VERSION 3.16)
project(mylibrary VERSION 1.0.0 LANGUAGES CXX)

add_library(mylibrary_core SHARED core.cpp)
add_library(mylibrary::core ALIAS mylibrary_core)
install(TARGETS mylibrary_core LIBRARY DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
```

A library alias is a constant reference to a library. An alias cannot be modified with linker or include options, and cannot be installed, because the alias name isn't necessarily a valid filename (e.g. it allows the colon symbol, used here to create a namespace).

One then uses `FetchContent(Mylibrary)` to load CMake targets `Mylibrary::core` and `Mylibrary_core` in the consumer project:

```cmake
cmake_minimum_required(VERSION 3.16)
project(myproject VERSION 4.3.0 LANGUAGES CXX)

include(FetchContent)
FetchContent_Declare(
  mylibrary
  GIT_REPOSITORY /path/to/mylibrary
)
if(NOT mylibrary_POPULATED)
  FetchContent_Populate(mylibrary)
  add_subdirectory("${mylibrary_SOURCE_DIR}" "${mylibrary_BINARY_DIR}")
endif()

add_library(myproject_core SHARED core.cpp)
add_library(myproject::core ALIAS myproject_core)
add_executable(myproject_main main.cpp)
add_executable(myproject::main ALIAS myproject_main)
target_link_libraries(myproject_main PRIVATE myproject::core mylibrary::core)

install(TARGETS myproject_core LIBRARY DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
install(TARGETS myproject_main RUNTIME DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
```

Importing the CMake target `Mylibrary_core` in the consumer project is actually desirable, because it gives the consumer project an opportunity to modify the library target with e.g. `target_link_libraries()`. The CMake target `Mylibrary::core` cannot be accidentally modified, since it is a constant reference to `Mylibrary_core`.

## Modern CMake using namespaced targets

Manually creating a target alias for every target may seem redundant considering that `target_link_libraries(myproject_main PRIVATE myproject_core)` leads to the same outcome as `target_link_libraries(myproject_main PRIVATE myproject::core)`. It may also be confusing to CMake non-experts, since one has to remember to always use the `myproject_main` when installing the target or adding source files / include directories / linker options to the target, and `myproject::main` when marking the target as a dependency of another target.

The systematic use of namespaced aliases comes with two benefits: it makes the use of `FetchContent` consistent with `find_package`, since both provide namespaces, and it leads to better diagnostics for missing dependencies. For example, consider a hypothetical project that includes a library `mylibrary` to get its `foo` component, but that component doesn't exist, either because `mylibrary` doesn't implement it, or it was not activated with CMake options, or it was not imported.

Option 1: using namespaced targets, CMake bails out immediately during configuration:

```cmake
target_link_libraries(myproject_main PRIVATE myproject::core mylibrary::core mylibrary::foo)
```
```
-- The CXX compiler identification is GNU 9.4.0
-- Configuring done
CMake Error at CMakeLists.txt:16 (add_executable):
  Target "myproject_main" links to target "mylibrary::foo" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?

-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly
```

Option 2: using real targets, CMake sees the string `mylibrary_foo` and tries to interpret it as a target, but since it doesn't exist, CMake falls back to interpreting the string as a system library `-lmylibrary_foo`.

```cmake
target_link_libraries(myproject_main PRIVATE myproject_core mylibrary_core mylibrary_foo)
```
```
-- The CXX compiler identification is GNU 9.4.0
-- Configuring done
-- Generating done
-- Build files have been written to: /work/jgrad/espresso/build
[...]
[100%] Linking CXX executable myproject_main
/usr/bin/ld: cannot find -lmylibrary_foo
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/myproject_main.dir/build.make:89: myproject_main] Error 1
make[1]: *** [CMakeFiles/Makefile2:102: CMakeFiles/myproject_main.dir/all] Error 2
make: *** [Makefile:133: all] Error 2
```

The configuration is successful and the build starts, only to fail much later at the linking stage. In this case, the linker complains `-lmylibrary_foo` was not found. But if there was a system-wide `libmylibrary_foo.so` file, the linker would have succeeded. This can be quite dangerous when the operating system already provides a different version of the same library via e.g. a standard Ubuntu package, because the linker would mix two different ABIs and potentially run into ODR violations.

There is one major objection to using namespaced aliases: if the project needs to be exported, all targets will be "namespaced twice", e.g. `mylibrary_core` will have to be accessed through `Mylibrary::mylibrary_core` in the consumer project when using `find_package(Mylibrary)` instead of `FetchContent()`. There have been several discussions about that issue in the CMake project (for a summary, see [cmake/cmake#22687](https://gitlab.kitware.com/cmake/cmake/-/issues/22687)), but at the moment, CMake projects that are willing to provide namespaced targets have to choose between two mutually incompatible solutions:

* exported targets in the library and `find_package()` in the consumer project, or
* target aliases in the library and `FetchContent()` in the consumer project.

While one could use a string replacement operation on the generated `FindMylibrary.cmake` to remove the double namespace, this would make the non-install version `export(EXPORT MylibraryTargets NAMESPACE Mylibrary::)` disagree with `FindMylibrary.cmake`.

## How is ESPResSo affected

ESPResSo currently uses a confusing mix of:
* prefixed target names:
  https://github.com/espressomd/espresso/blob/ef9f4979ee3ab753f44e803147fb9ae75ace2ffc/src/core/CMakeLists.txt#L67
  https://github.com/espressomd/espresso/blob/ef9f4979ee3ab753f44e803147fb9ae75ace2ffc/src/utils/CMakeLists.txt#L1

* non-prefixed target names:
  https://github.com/espressomd/espresso/blob/ef9f4979ee3ab753f44e803147fb9ae75ace2ffc/src/core/cluster_analysis/CMakeLists.txt#L2
  https://github.com/espressomd/espresso/blob/ef9f4979ee3ab753f44e803147fb9ae75ace2ffc/src/profiler/CMakeLists.txt#L1
  https://github.com/espressomd/espresso/blob/ef9f4979ee3ab753f44e803147fb9ae75ace2ffc/src/script_interface/CMakeLists.txt#L1-L2
  https://github.com/espressomd/espresso/blob/ef9f4979ee3ab753f44e803147fb9ae75ace2ffc/CMakeLists.txt#L344

The non-prefixed targets are prone to collisions with waLBerla and any other library we may need in the future. To make things worse, some of the non-prefixed targets (e.g. `Profiler`) are opt-in, which leads to cryptic linker errors if they are included as dependencies of other targets without a proper guard. I actually spent some time investigating what package provided `-lProfiler` during a CMake refactor, until I realized it was an ESPResSo target missing the `Espresso` prefix... Using namespaced aliases would make our CMake logic more maintainable by producing more meaningful error messages, as well as reducing the risk of accidentally modifying a target with a copy-paste error (namespaced aliases are immutable).

To solve both the issue of name collision and the issue of missing targets being interpreted as strings in the waLBerla dependency, the `walberla` branch of ESPResSo introduced the following patch to automatically generate a namespaced alias for all waLBerla targets:

https://github.com/espressomd/espresso/blob/231816491ffb7922bd1918e74069b46068201a58/cmake/waLBerlaFunctions.patch#L1-L7

We could apply the same treatment to ESPResSo. A possible implementation could look like this: jngrad/espresso@778f12526e41.
